<div style="text-align: center;">
    <img src="https://upload.wikimedia.org/wikipedia/commons/thumb/f/f8/Python_logo_and_wordmark.svg/972px-Python_logo_and_wordmark.svg.png?20210516005643" alt="The Python logo" style="width: 60%; max-width: 500px;">
</div>

<div class="alert alert-success" role="alert">

<div style="display: flex; justify-content: space-between;">
<div style="width: 48%; line-height: 1.5;">

### Instructions

- **Enter your matriculation number in the first cell at the line `student_id =` and then execute the cell.**

- Modify this notebook directly.

- <span style="color:red;"> Do not clear the outputs of your cells!</span>

- **Import any necessary modules.**

- Insert additional cells as needed after each task.

- **Add comments to your code**, either directly in the Python cell or using separate Markdown cells.

- Once completed, upload your notebook to the weekly assignment.

- **There are 11 points available in total: you need at least 7 points to pass.**

</div>
<div style="width: 4%;">
</div>
<div style="width: 48%; line-height: 1.5; color: grey;">

### Anweisungen

- **Geben Sie Ihre Matrikelnummer in die erste Zelle der Zeile `student_id =` ein und führen Sie dann die Zelle aus.**

- Ändern Sie dieses Notizbuch direkt.

- <span style="color:red;"> Löschen Sie nicht die Ausgaben Ihrer Zellen!</span>

- **Importieren Sie alle notwendigen Module.**

- Fügen Sie bei Bedarf nach jeder Aufgabe zusätzliche Zellen ein.

- **Fügen Sie Kommentare zu Ihrem Code** hinzu, entweder direkt in der Python-Zelle oder in separaten Markdown-Zellen.

- Wenn Sie fertig sind, laden Sie Ihr Notizbuch in die wöchentliche Aufgabe hoch.

- **Es gibt insgesamt 11 Punkte: Sie brauchen mindestens 7 Punkte, um zu bestehen.**

</div>
</div>

In [ ]:
from IPython.display import display
import IPython
from datetime import datetime

# Student ID
student_id = 

def display_student_id(*args, **kwargs):
    current_time = datetime.now().strftime("%Y-%m-%d %H:%M:%S")
    display(f"Run by {student_id} at {current_time}")

IPython.get_ipython().events.register('post_run_cell', display_student_id)

<div class="alert alert-block alert-light">

#### 1 (1 P)

<div style="display: flex; justify-content: space-between;">
<div style="width: 48%; line-height: 1.5;">

The code below attempts to calculate the balance of a savings account at the end of each year for a maximum of four years.  The account starts with a balance of 100 EUR, and the annual interest rate is 0.72%.  

However, the code is not functioning as expected. Identify the issue and offer a corrected version.

</div>
<div style="width: 48%; line-height: 1.5;color: grey;">

Der folgende Code versucht, den Saldo eines Sparkontos am Ende eines jeden Jahres für maximal vier Jahre zu berechnen.  Das Konto beginnt mit einem Saldo von 100 EUR, und der jährliche Zinssatz beträgt 0.72%.  

Der Code funktioniert jedoch nicht wie erwartet. Ermitteln Sie den Fehler und bieten Sie eine korrigierte Version an.
</div>
</div>

```python
# The following code does not work as intended - identify the problem and correct it
balance = 100

def add_interest(balance, rate):
    balance += balance * rate/100
    
for year in range(1, 5):
    add_interest(balance, 0.72)
    print ('Balance after year', year, '=', balance)
```

<div style="display: flex; justify-content: space-between;">
<div style="width: 48%; line-height: 1.5;">

The code fails because the `balance` variable inside the function is treated as local, creating a separate variable that does not affect the global `balance`. To fix this, the function must return the updated value, and you need to reassign it outside the function.

You can modify the function to return the updated balance and then capture the returned value in the main code loop.

- The function returns the updated `balance`, which is then reassigned to the global `balance` variable after each year.
- This way, the global `balance` is updated correctly, and the output reflects the change.

</div>
<div style="width: 48%; line-height: 1.5;color: grey;">

Der Code schlägt fehl, weil die Variable `balance` innerhalb der Funktion als lokal behandelt wird, wodurch eine separate Variable entsteht, die die globale Variable `balance` nicht beeinflusst. Um dies zu beheben, muss die Funktion den aktualisierten Wert zurückgeben, und Sie müssen ihn außerhalb der Funktion neu zuweisen.

Sie können die Funktion so ändern, dass sie den aktualisierten Saldo zurückgibt, und den zurückgegebenen Wert dann in der Hauptcodeschleife erfassen.

- Die Funktion gibt den aktualisierten `balance` zurück, der dann nach jedem Jahr der globalen Variable `balance` neu zugewiesen wird.
- Auf diese Weise wird der globale `balance` korrekt aktualisiert, und die Ausgabe spiegelt die Änderung wider.

</div>
</div>

In [17]:
balance = 100

def add_interest(balance, rate):
    balance += balance * rate / 100
    return balance  # Return the updated balance

# Now capture the returned value in the loop
for year in range(1, 5):
    balance = add_interest(balance, 0.72)  # Update balance with the returned value
    print('Balance after year', year, '=', balance)

Balance after year 1 = 100.72
Balance after year 2 = 101.445184
Balance after year 3 = 102.1755893248
Balance after year 4 = 102.91125356793856


<div class="alert alert-block alert-light">

#### 2 (2 P)

<div style="display: flex; justify-content: space-between;">
<div style="width: 48%; line-height: 1.5;">

The **first-principles definition** of a derivative:
</div>
<div style="width: 48%; line-height: 1.5;color: grey;">

der **Erstprinzipien-Definition** einer Ableitung:
</div>
</div>
<div style="text-align: center; margin-top: 20px;">

$$
f'(x) = \lim_{h \to 0} \frac{f(x+h) - f(x)}{h}
$$
</div>
<div style="display: flex; justify-content: space-between;">
<div style="width: 48%; line-height: 1.5;">

shows that to compute a derivative numerically, we only need the definition of the function $f(x)$ and a value for the step size.

Here, $f(x)$ is the original function, $f'(x)$ is its derivative, and $h$ is the step size. 

Calculate the derivative of $f(x) = x^4$ at $x = 1.3$. 

The analytical solution, $f'(x) = 4x^3$, gives a result of $8.788$. 

1. Define the mathematical function $f(x)=x^{4}$ using python.

2. Using the first principles definition of a derivative, write a second function, `derivative()`, containing two parameters: the stepsize, $h$ (which we shall give a default value of $0.01$), and the value of $x$ we are using to evaluate the derivative.

**Useful tips:**

- Functions can be called from within other functions, and their output used as normal.

- Both `x` and `h` are numbers, therefore `x+h` is also a number, and can be passed to a function as an argument.
</div>
<div style="width: 48%; line-height: 1.5;color: grey;">

zeigt, dass wir für die numerische Berechnung einer Ableitung nur die Definition der Funktion $f(x)$ und einen Wert für die Schrittweite benötigen.

$f(x)$ ist die Funktion, $f'(x)$ ist ihre Ableitung und $h$ ist die Schrittweite. 

Berechnen Sie die Ableitung von $f(x) = x^4$ bei $x = 1.3$. 

Die analytische Lösung, $f'(x) = 4x^3$, liefert ein Ergebnis von $8.788$.

1. Definieren Sie die mathematische Funktion $f(x)=x^{4}$ mit Python.

2. Schreiben Sie unter Verwendung der Definition einer Ableitung nach den ersten Prinzipien eine zweite Funktion, `derivative()`, die zwei Parameter enthält: die Schrittweite $h$ (die wir mit $0.01$ vorgeben) und den Wert von $x$, den wir zur Auswertung der Ableitung verwenden.

**Nützliche Tipps:**

- Funktionen können aus anderen Funktionen heraus aufgerufen werden, und ihre Ausgabe wird wie üblich verwendet.

- Sowohl `x` als auch `h` sind Zahlen, daher ist `x+h` auch eine Zahl und kann als Argument an eine Funktion übergeben werden.
</div>
</div>

In [2]:
def f(x):
    y = x**4
    return y

f(8)

4096

In [3]:
def derivative(x, h=0.01):
    f_prime = (f(x+h) - f(x)) / h
    return f_prime

derivative(1.3, 0.01)

8.889921000000012

<div class="alert alert-block alert-light">

#### 3  (2 P)

<div style="display: flex; justify-content: space-between;">
<div style="width: 48%; line-height: 1.5;">

Your `derivative()` function should have returned a value of approximately 8.89 for the derivative of $x^4$ evaluated at $x=1.3$. 
While this is close to the exact value of $8.788$, can we improve the accuracy?

The first-principles definition of a derivative assumes the limit as $h$ approaches zero. 
This implies that by decreasing $h$ (getting closer to $h = 0$), our numerical approximation should become more precise. 

Using your `derivative()` function, repeat the calculation above with step sizes of 
$$ h = 1, 0.1, 0.01, 0.001, 0.0001, 0.00001.$$ 

Use a loop in your code to automate the function calls, instead of manually invoking the function for each step size.

Print the results as an f-string table.

</div>
<div style="width: 48%; line-height: 1.5;color: grey;">

Ihre Funktion `derivative()` sollte für die Ableitung von $x^4$ nach $x=1.3$ einen Wert von etwa 8.89 liefern. 
Dies liegt zwar nahe am exakten Wert von $8.788$, aber können wir die Genauigkeit verbessern?

Die Definition einer Ableitung nach den ersten Grundsätzen geht davon aus, dass der Grenzwert bei $h$ gegen Null geht. 
Das bedeutet, dass unsere numerische Näherung mit abnehmendem $h$ (näher an $h = 0$) präziser werden sollte.

Wiederholen Sie die obige Berechnung mit der Funktion `derivative()` mit Schrittweiten von
$$ h = 1, 0.1, 0.01, 0.001, 0.0001, 0.00001.$$ 

Verwenden Sie eine Schleife in Ihrem Code, um die Funktionsaufrufe zu automatisieren, anstatt die Funktion für jede Schrittweite manuell aufzurufen.

Geben Sie die Ergebnisse als f-String-Tabelle aus.

</div>
</div>

In [8]:
step_sizes = [1, 0.1, 0.01, 0.001, 0.0001, 0.00001]

for h in step_sizes:
    print(f'{h:.0e}\t{derivative(1.3, h):>10.6f}')

1e+00	 25.128000
1e-01	  9.855000
1e-02	  8.889921
1e-03	  8.798145
1e-04	  8.789014
1e-05	  8.788101


<div class="alert alert-block alert-light">

#### 4 (1 P)

<div style="display: flex; justify-content: space-between;">
<div style="width: 48%; line-height: 1.5;">

Write a program that can compute the derivative with respect to $x$ for any mathematical function of the form:
$f(x) = ax^p$.

Test your program by evaluating the derivative of $f(x) = 4x^3$ at $x = 3$. 

If your code is working correctly, it should return a value close to $108$ (the exact value may vary slightly depending on the step size chosen).
</div>
<div style="width: 48%; line-height: 1.5;color: grey;">

Schreiben Sie ein Programm, das die Ableitung nach $x$ für eine beliebige mathematische Funktion der Form berechnen kann:
$f(x) = ax^p$.

Testen Sie Ihr Programm, indem Sie die Ableitung von $f(x) = 4x^3$ nach $x = 3$ auswerten. 

Wenn Ihr Programm korrekt arbeitet, sollte es einen Wert nahe $108$ zurückgeben (der genaue Wert kann je nach der gewählten Schrittweite leicht variieren).
</div>
</div>

In [23]:
def f(x, a, p):
    y = 4*x**p
    return y

def derivative(x, a, p, h):
    value = (f(x+h, a, p) - f(x, a, p)) / h
    print(value)
    
derivative(x=3, a=4, p=3, h=1.e-6)

108.00003602184916


<div class="alert alert-block alert-light">

#### 5 (2 P)

<div style="display: flex; justify-content: space-between;">
<div style="width: 48%; line-height: 1.5;">

Write a Python function to calculate the heat capacity according to the Einstein model given below.

Then, calculate the heat capacities of sodium and diamond at 300 K, given that $\Theta_E$ is 192 K for sodium and 1450 K for diamond.

</div>
<div style="width: 48%; line-height: 1.5;color: grey;">

Schreiben Sie eine Python-Funktion zur Berechnung der Wärmekapazität nach dem unten angegebenen Einstein-Modell.

Berechnen Sie dann die Wärmekapazitäten von Natrium und Diamant bei 300 K, wobei $\Theta_E$ für Natrium 192 K und für Diamant 1450 K beträgt.

</div>
</div>
<div style="text-align: center; margin-top: 20px;">

$$
C_{V,m} = 3R \left( \frac{\Theta_E}{T} \right)^2 \frac{\exp\left( \frac{\Theta_E}{T} \right)}{\left[ \exp\left( \frac{\Theta_E}{T} \right) - 1 \right]^2}
$$
</div>
<div style="display: flex; justify-content: space-between;">
<div style="width: 48%; line-height: 1.5;">

**Hints**:

- Remember to import all relevant modules (and constants).

- Define a variable $\Theta_E / T$ for your convenience.

</div>
<div style="width: 48%; line-height: 1.5;color: grey;">

**Hinweise**:

- Denken Sie daran, alle relevanten Module (und Konstanten) zu importieren.

- Definieren Sie eine Variable $\Theta_E / T$ für Ihre Bequemlichkeit.

</div>
</div>

In [27]:
from scipy.constants import R
import math

def C_V(T, Theta_E):
    x = Theta_E / T
    return 3 * R * x**2 * math.exp(x) / (math.exp(x) - 1)**2

print(f"Na: {C_V(300, 192)}")
print(f"diamond: {C_V(300, 1450)}")

Na: 24.109144168075208
diamond: 4.713033453464986


<div class="alert alert-block alert-light">

#### 6 (3 P)

<div style="display: flex; justify-content: space-between;">
<div style="width: 48%; line-height: 1.5;">

For the following equations of state, define Python functions returning the pressure of a gas, $p$, given its volume $V$, amount $n$, and temperature $T$.

- **Ideal gas equation of state**: $$ p = \frac{nRT}{V},$$
- **Van der Waals equation of state**: $$ p = \frac{nRT}{V-nb} - \frac{n^2a}{V^2},$$
- **Dieterici equation of state**: $$ p = \frac{nRT \exp \left( -\frac{nc}{RTV} \right)}{V-nd}.$$

Use the following parameters:

- $T = 273.15 \, \text{K}$;
- $V = 20 \, \text{L}$;
- $a = 3.640 \, \text{L}^2 \, \text{bar} \, \text{mol}^{-2}, \quad b = 0.04267 \, \text{L} \, \text{mol}^{-1}$;
- $c = 4.692 \, \text{L}^2 \, \text{bar} \, \text{mol}^{-2}, \quad d = 0.04639 \, \text{L} \, \text{mol}^{-1}$;
- $n = 1 \, \text{mol}$.

Repeat the calculation for $n = 0.1 \, \text{mol}$.

Print summaries using string formatting.
</div>
<div style="width: 48%; line-height: 1.5;color: grey;">

Definieren Sie für die folgenden Zustandsgleichungen Python-Funktionen, die den Druck eines Gases $p$ wiedergeben, wenn das Volumen $V$, die Menge $n$ und die Temperatur $T$ gegeben sind.

- **Ideale Gaszustandsgleichung**: $$ p = \frac{nRT}{V},$$
- **Van-der-Waals-Zustandsgleichung**: $$ p = \frac{nRT}{V-nb} - \frac{n^2a}{V^2},$$
- **Dieterici-Zustandsgleichung**: $$ p = \frac{nRT \exp \left( -\frac{nc}{RTV} \right)}{V-nd}.$$

Verwenden Sie die folgenden Parameter:

- $T = 273.15 \, \text{K}$;
- $V = 20 \, \text{L}$;
- $a = 3.640 \, \text{L}^2 \, \text{bar} \, \text{mol}^{-2}, \quad b = 0.04267 \, \text{L} \, \text{mol}^{-1}$;
- $c = 4.692 \, \text{L}^2 \, \text{bar} \, \text{mol}^{-2}, \quad d = 0.04639 \, \text{L} \, \text{mol}^{-1}$;
- $n = 1 \, \text{mol}$.

Wiederholen Sie die Berechnung für $n = 0.1 \, \text{mol}$.

Zusammenfassungen unter Verwendung der String-Formatierung drucken.
</div>
</div>